app/src/main/java/com/wireguard/android/databinding/ObservableKeyedRecyclerViewAdapter.kt

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright © 2017-2018 WireGuard LLC. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.wireguard.android.databinding
+
+import android.content.Context
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.databinding.DataBindingUtil
+import androidx.databinding.ObservableList
+import androidx.databinding.ViewDataBinding
+import androidx.recyclerview.widget.RecyclerView
+import androidx.recyclerview.widget.RecyclerView.Adapter
+import com.wireguard.android.BR
+import com.wireguard.android.util.ObservableKeyedList
+import com.wireguard.util.Keyed
+
+import java.lang.ref.WeakReference
+
+/**
+ * A generic `RecyclerView.Adapter` backed by a `ObservableKeyedList`.
+ */
+
+class ObservableKeyedRecyclerViewAdapter<K, E : Keyed<out K>> internal constructor(
+    context: Context, private val layoutId: Int,
+    list: ObservableKeyedList<K, E>
+) : Adapter<ObservableKeyedRecyclerViewAdapter.ViewHolder>() {
+
+    private val callback = OnListChangedCallback(this)
+    private val layoutInflater: LayoutInflater = LayoutInflater.from(context)
+    private var list: ObservableKeyedList<K, E>? = null
+    private var rowConfigurationHandler: RowConfigurationHandler<ViewDataBinding,E>? = null
+
+    init {
+        setList(list)
+    }
+
+    override fun getItemCount(): Int {
+        return if (list != null) list!!.size else 0
+    }
+
+    private fun getItem(position: Int): E? {
+        return if (list == null || position < 0 || position >= list!!.size) null else list!![position]
+    }
+
+    override fun getItemId(position: Int): Long {
+        val key = getKey(position)
+        return (key?.hashCode() ?: -1).toLong()
+    }
+
+    private fun getKey(position: Int): K? {
+        val item = getItem(position)
+        return item?.key
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        return ViewHolder(DataBindingUtil.inflate(layoutInflater, layoutId, parent, false))
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.binding.setVariable(BR.collection, list)
+        holder.binding.setVariable(BR.key, getKey(position))
+        holder.binding.setVariable(BR.item, getItem(position))
+        holder.binding.executePendingBindings()
+
+        val item = getItem(position)
+        if (item != null) {
+            rowConfigurationHandler?.onConfigureRow(holder.binding, item, position)
+        }
+    }
+
+    internal fun setList(newList: ObservableKeyedList<K, E>?) {
+        if (list != null)
+            list!!.removeOnListChangedCallback(callback)
+        list = newList
+        if (list != null) {
+            list!!.addOnListChangedCallback(callback)
```

Comment:
Ditto

Reply:
I didn't get functionality . So I thought it should report for null . Will do it

Reply:
`?.` is simply a null-check wrapped into the actual statement. If the object on the left side of `?.` is null, the statement is never executed.

Reply:
I thought this would be important statement. That's why explicitly  !! To throw null, so that would get to know.  Will take care of this then